src/linear_algebra/matrix/symmetric.lean

```diff
@@ -0,0 +1,118 @@
+/-
+Copyright (c) 2021 Lu-Ming Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Lu-Ming Zhang
+-/
+import data.matrix.block
+
+/-!
+# Symmetric matrices
+
+This file contains the definition and basic results about symmetric matrices.
+
+## Main definition
+
+ * `matrix.is_sym `: a matrix `A : matrix n n α` is "symmetric" if `Aᵀ = A`.
+
+## Tags
+
+sym, symmetric, matrix
+-/
+
+variables {α n m R : Type*}
+
+namespace matrix
+
+open_locale matrix
+
+/-- A matrix `A : matrix n n α` is "symmetric" if `Aᵀ = A`. -/
+def is_sym (A : matrix n n α) : Prop := Aᵀ = A
+
+lemma is_sym.eq {A : matrix n n α} (h : A.is_sym) : Aᵀ = A := h
+
+lemma is_sym.ext_iff {A : matrix n n α} : A.is_sym ↔ ∀ i j, Aᵀ i j = A i j :=
+by rw [is_sym, matrix.ext_iff]
+
+lemma is_sym.ext_iff' {A : matrix n n α} : A.is_sym ↔ ∀ i j, A j i = A i j :=
+is_sym.ext_iff
+
+lemma is_sym.apply {A : matrix n n α} (h : A.is_sym) (i j : n) : Aᵀ i j = A i j :=
+is_sym.ext_iff.1 h i j
+
+lemma is_sym.apply' {A : matrix n n α} (h : A.is_sym) (i j : n) : A j i = A i j :=
+is_sym.apply h i j
+
+lemma is_sym.ext {A : matrix n n α} : (∀ i j, Aᵀ i j = A i j) → A.is_sym :=
+is_sym.ext_iff.2
+
+lemma is_sym.ext' {A : matrix n n α} : (∀ i j, A j i = A i j) → A.is_sym :=
+is_sym.ext
+
+/-- A block matrix `A.from_blocks B C D` is symmetric,
+    if `A` and `D` are symmetric and `Bᵀ = C`. -/
+lemma is_sym_from_blocks
+  {A : matrix m m α} {B : matrix m n α} {C : matrix n m α} {D : matrix n n α}
+  (h1 : A.is_sym) (h2 : D.is_sym) (h3 : Bᵀ = C) :
+  (A.from_blocks B C D).is_sym :=
+begin
+  have h4 : Cᵀ = B, {rw ← h3, simp},
+  unfold matrix.is_sym,
+  rw from_blocks_transpose,
+  congr;
+  assumption
+end
+
+/-- `A ⬝ Aᵀ` is symmertric. -/
+lemma is_sym_mul_transpose_self [fintype n] [comm_semiring α] (A : matrix n n α) :
+  (A ⬝ Aᵀ).is_sym :=
```

Comment:
Can you add a copy of this for when the multiplication is in the other direction?

Reply:
And also for addition